factory.py

```diff
@@ -64,14 +83,22 @@ def __pre_init__(self, *args, **kwargs):
         """Pre-initialization method. Use/extend this method to perform any checks before initializing the class and
         fail fast to adapt to other factory classes."""
         kw = args[0]
-        assert (
-            FactoryTypes(kw.get(FactoryArgumentKeys.TASK, None)) == self._type
-        ), "Invalid task type."
 
-        algorithm_name = kw.get(FactoryArgumentKeys.ALGORITHM, None)
-        assert (
-            self.get_algorithm(algorithm_name) is not None
-        ), f"Invalid algorithm: {algorithm_name}"
+        if kw.get(FactoryArgumentKeys.JOB_SERVICE, None) == JobServices.ADAPT.value:
+            assert (
+                FactoryTypes(kw.get(FactoryArgumentKeys.JOB_SERVICE, None))
+                == self._type
+            ), "Invalid task type."
+
+        else:
+            assert (
+                FactoryTypes(kw.get(FactoryArgumentKeys.TASK, None)) == self._type
+            ), "Invalid task type."
+
+            algorithm_name = kw.get(FactoryArgumentKeys.ALGORITHM, None)
+            assert (
+                self.get_algorithm(algorithm_name) is not None
+            ), f"Invalid algorithm: {algorithm_name}"
```

Comment:
if the _pre_init_ is defined here, why not have it in one palce for all modules? What is the issue with conditionals here ? This is supposed to fail fast for every service, so why not have it here?

Reply:
the purpose of extending a method is to avoid integrational / module level conditionals. since we have diff factories for diff method, each factory is expected to implement its own pre init. The only reason kompress didn't have it implemented is because these  assertions were more than sufficient for the both the factories and must be sufficient for a few primary fail fasts for adapt as well - such as get.algorithm and factory type; 
now, we can expect some changes to the factory type assertion since we have two arguments to the factory (service and task) but other than that I dont see major changes in the print

Reply:
I don't see any point in changing for a single conditional statement here right ?

Reply:
we do it for separation of logic. besides, we must not need any major change to assertions here except for maybe name changes or parameter changes. as adapt must assert a get_algorithm, and factory type at the very least.

Reply:
adapt doesn't have any get_algorithm.  So, I don't think it needs to assert that.  And we are asserting the factory type here from JOB_SERVICE right ?

Reply:
the default factory flow expects a get_algorithm method to fail fast if the algorithm is not available in a factory type. if u have a different method in adapt that does this, just call that method in adapt's get_algorithm method. or set a variable in adapt and return that variable when this method is called.